zebra-chain/src/block/difficulty.rs

```diff
@@ -28,18 +37,69 @@ use proptest_derive::Arbitrary;
 /// an 8-bit exponent, an offset of 3, and a radix of 256.
 /// (IEEE 754 32-bit floating-point values use a separate sign bit, an implicit
 /// leading mantissa bit, an offset of 127, and a radix of 2.)
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
+///
+/// The precise bit pattern of a `CompactDifficulty` value is
+/// consensus-critical, because it is used for the `difficulty_threshold` field,
+/// which is:
+///   - part of the `BlockHeader`, which is used to create the
+///     `BlockHeaderHash`, and
+///   - bitwise equal to the median `ExpandedDifficulty` value of recent blocks,
+///     when encoded to `CompactDifficulty` using the specified conversion
+///     function.
+///
+/// Without these consensus rules, some `ExpandedDifficulty` values would have
+/// multiple equivalent `CompactDifficulty` values, due to redundancy in the
+/// floating-point format.
+#[derive(Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
 #[cfg_attr(test, derive(Arbitrary))]
 pub struct CompactDifficulty(pub u32);
 
-/// A 256-bit expanded difficulty value.
+impl fmt::Debug for CompactDifficulty {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("CompactDifficulty")
+            // Use hex, because it's a float
+            .field(&format_args!("{:#010x}", self.0))
+            .finish()
+    }
+}
+
+/// A 256-bit unsigned "expanded difficulty" value.
 ///
 /// Used as a target threshold for the difficulty of a `BlockHeaderHash`.
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
-#[cfg_attr(test, derive(Arbitrary))]
-pub struct ExpandedDifficulty([u8; 32]);
+///
+/// Details:
+///
+/// The precise bit pattern of an `ExpandedDifficulty` value is
+/// consensus-critical, because it is compared with the `BlockHeaderHash`.
+///
+/// Note that each `CompactDifficulty` value represents a range of
+/// `ExpandedDifficulty` values, because the precision of the
+/// floating-point format requires rounding on conversion.
+///
+/// Therefore, consensus-critical code must perform the specified
+/// conversions to `CompactDifficulty`, even if the original
+/// `ExpandedDifficulty` values are known.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct ExpandedDifficulty(U256);
```

Comment:
Mostly ok with pulling in this type, I kinda wish there was a generic std Rust bigint like this, @yaahc

Reply:
Yeah I wish we had a generic `u256`, but I looked around, and I couldn't find a better one.

If there is a `u256` in future, it should be easy to replace. I'm not using any custom methods on that type.